### ISSUE TITLE
Differing fetchSteam hashes with different nixpkgs versions

### ISSUE THREAD

**Issue by olistrik** [aidalgol/valheim-server-flake#7]
I'm not sure if this an issue better addressed here, or upstream on steam-fetcher, eitherway I've noticed that I get different hashes for the `steamFetch` on `nixpkgs-23.05` and `nixpkgs-23.11`. There's a good chance that I'm just miss using or abusing something, but I thought I'd ask anyway.

I understand that if the upstream DepotDownloader changes how it fetches steamapps then the hashes may change, though I do find it odd as the src itself hasn't changed. The problem for me is that even if I leave the nixpkgs input of this flake alone, when overlaying it onto `nixpkgs-23.11` it no-longer builds itself with `nixpkgs-23.05`. 

Given how overlays are a function which take the `final` and `prev` args of the `nixpkgs` that they're overlaying, this behavior is hardly surprising. But it makes it difficult to use this flake with any nixpkgs other than the set it's locked too.

I could update hash and the flake inputs but then it'll break for anyone on `23.05`.

I'm using snowfall-lib, so this config is paraphrased:

```nix
{
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-23.11";
    valheim-server = {
      # no follows; so nixpkgs = 23.05.
      url = "github:aidalgol/valheim-server-flake";
    };
  };
  outputs = inputs: 
    let
      lib = // ...;
    in
    lib.mkFlake {
      //...
      overlays = with inputs; [
        valheim-server.overlays.default
      ];

      systems.modules.nixos = with inputs; [
        valheim-server.nixosModules.default
      ];
    };
 }
```

**Comment by aidalgol** [comment]
That is not something I had even considered.  I suppose since DepotDownloader is an unofficial hack, I should not be surprised that a later version may produce a slightly different result.

I've pushed a fix to the Steam fetcher to try the version of DepotDownloader to the nixpkgs input in the `steam-fetcher` flake, but I'm not able to easily test it.  Could you try pointing your NixOS config flake's `valheim-server` input to the `fetch-hash-issue` branch and see if that can build?

**Comment by olistrik** [comment]
@aidalgol it didn't fix it unfortunately. `prev` is still whatever package set that is being overlayed. However it did give me an idea on how to fix it. I've made two MRs:

- [one on steam-fetcher](https://github.com/nix-community/steam-fetcher/pull/3) that pins depotdownloader to the flakes package set for the fetchSteam derivation. 
- [the other here](https://github.com/aidalgol/valheim-server-flake/pull/8) pinning the fetchSteam fetcher to the flakes own steam-fetcher input for the valheim-server derivation.

I have both `steam-fetcher` and `valheim-server-flake` as inputs, and I've found this approach gives the most flexibility, there's a few different combinations of follows but I think their results make the most sense.

basically, if you don't set nixpkgs.follows, they'll use their own nixpkgs input for the fetchSteam/depotdownloader in the overlays.

if you set the maximum follows:

```
nixpkgs.url = "23.11" 
steam-fetcher.nixpkgs.follows = "nixpkgs"
valheim-server.nixpkgs.follows = "nixpkgs"
valheim-server.steam-fetcher.follows = "steam-fetcher"
```

then you'll only have one revision of nixpkgs, steam-fetcher, and valheim server, and all of them will use `23.11`.

**Comment by tie** [comment]
@aidalgol FWIW the merged PR probably breaks cross-compilation because pkgs.depotdownloader is not spliced, and final.system is stdenv.hostPlatform.system (note that system attribute in general _is not_ recommended for use), and also breaks overlays for depotdownloader and other dependencies. The implementation seems like non-idiomatic Nix(pkgs) to me.

Hashes are probably different because DepotDownloader leaves some metadata around in the output directory. See my implementation for reference:
https://github.com/tie-infra/steam-games/blob/989e86dda4aa7ecd6261ad2e595a1d43190ded07/fetch-steam/pkgs/fetch-steam/default.nix#L55

Edit: DepotDownloader is using [protobufs (with deflate)](https://github.com/SteamRE/DepotDownloader/blob/b96125f9cbbb0f63d47e14784929f255f6c21ce1/DepotDownloader/DepotConfigStore.cs) for serialization, and protobuf is infamous for being [Deterministic but not Canonical](https://protobuf.dev/programming-guides/serialization-not-canonical/#deterministic-is-not-canonical).
E.g. for Go protobuf serialization is deterministically random, see [google.golang.org/protobuf/internal/detrand](https://pkg.go.dev/google.golang.org/protobuf/internal/detrand). I assume that is also the case for C#.
>The pseudo-randomness of these functions is seeded by the program binary itself and guarantees that the output does not change within a program, while ensuring that the output is unstable across different builds.

Anyway, the fix is to just `rm -r` metadata directory.

**Comment by aidalgol** [comment]
> @aidalgol FWIW the merged PR probably breaks cross-compilation because pkgs.depotdownloader is not spliced, and final.system is stdenv.hostPlatform.system (note that system attribute in general _is not_ recommended for use), and also breaks overlays for depotdownloader and other dependencies. The implementation seems like non-idiomatic Nix(pkgs) to me.

Thank you very much for pointing this out.  I really should have inspected what DepotDownloader was doing more closely, and then I may have found the `.DepotDownloader` directory.

> Edit: DepotDownloader is using [protobufs (with deflate)](https://github.com/SteamRE/DepotDownloader/blob/b96125f9cbbb0f63d47e14784929f255f6c21ce1/DepotDownloader/DepotConfigStore.cs) for serialization, and protobuf is infamous for being [Deterministic but not Canonical](https://protobuf.dev/programming-guides/serialization-not-canonical/#deterministic-is-not-canonical).
E.g. for Go protobuf serialization is deterministically random, see [google.golang.org/protobuf/internal/detrand](https://pkg.go.dev/google.golang.org/protobuf/internal/detrand). I assume that is also the case for C#.

Oh no... does this mean the output will still change across versions, or does this only affect the contents of that metadata directory?

**Comment by tie** [comment]
Should be only `.DepotDownloader` directory, yes.

**Comment by aidalgol** [comment]
OK, great.  Then this should be fixed as of baced068120236bcf4f4552ff560df0624a689e3.  (See also latest changes on [nix-community/steam-fetcher](https://github.com/nix-community/steam-fetcher).)
